node/src/services/blockchain/transaction.rs

```diff
@@ -0,0 +1,150 @@
+            let remaining = match self.timeout {
+                Some(timeout) => {
+                    // Check if the timeout has been reached.
+                    if elapsed > timeout {
+                        return Err(anyhow!(
+                            "Timeout waiting for transaction to be included in a block"
+                        ));
+                    }
+
+                    timeout - elapsed
+                }
+                None => NO_TIMEOUT_WARN,
+            };
+
+            // Wait for either a new message from the watcher, or the timeout to be reached.
+            let result = match tokio::time::timeout(remaining, self.watcher.recv()).await {
+                Ok(result) => match result {
+                    Some(result) => result,
+                    None => {
+                        return Err(anyhow!("Transaction watcher channel closed"));
+                    }
+                },
+                Err(_) => {
+                    // Timeout reached, exit the loop.
+                    match self.timeout {
+                        Some(_) => {
+                            return Err(anyhow!(
+                                "Timeout waiting for transaction to be included in a block"
+                            ));
+                        }
+                        None => {
+                            // No timeout set, continue waiting.
+                            warn!(target: LOG_TARGET, "No timeout set and {:?} elapsed, continuing to wait for transaction to be included in a block.", NO_TIMEOUT_WARN);
+
+                            continue;
+                        }
+                    }
+                }
+            };
+            // Parse the JSONRPC string, now that we know it is not an error.
+            let json: serde_json::Value = serde_json::from_str(&result).map_err(|_| {
+                anyhow!("The result, if not an error, can only be a JSONRPC string; qed")
+            })?;
+
+            debug!(target: LOG_TARGET, "Transaction information: {:?}", json);
+
+            // Checking if the transaction is included in a block.
+            // TODO: Consider if we might want to wait for "finalized".
+            // TODO: Handle other lifetime extrinsic edge cases. See https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/transaction-pool/api/src/lib.rs#L131
+            if let Some(in_block) = json["params"]["result"]["inBlock"].as_str() {
+                block_hash = Some(H256::from_str(in_block)?);
+                let subscription_id = json["params"]["subscription"]
+                    .as_number()
+                    .ok_or_else(|| anyhow!("Subscription should exist and be a number; qed"))?;
+
+                // Unwatch extrinsic to release tx_watcher.
+                blockchain
+                    .unwatch_extrinsic(subscription_id.to_owned())
+                    .await?;
+
+                // Breaking while loop.
```

Comment:
```suggestion
                // Breaking loop.
```

Reply:
It is actually a `loop` not a `while` anymore.